app/controlplane/pkg/ca/ejbca/ejbca.go

```diff
@@ -0,0 +1,170 @@
+	SerialNumber     string   `json:"serial_number,omitempty"`
+	ResponseFormat   string   `json:"response_format,omitempty"`
+	CertificateChain []string `json:"certificate_chain,omitempty"`
+}
+
+func (e EJBCA) CreateCertificateFromCSR(ctx context.Context, principal identity.Principal, csr *x509.CertificateRequest) (*fulcioca.CodeSigningCertificate, error) {
+	ejbcaReq := &Request{
+		CertificateRequest:       string(csr.Raw),
+		CertificateProfileName:   e.certificateProfileName,
+		EndEntityProfileName:     e.endEntityProfileName,
+		CertificateAuthorityName: e.caName,
+		Username:                 principal.Name(ctx),
+		IncludeChain:             true,
+	}
+	ejbcaReqBytes, err := json.Marshal(ejbcaReq)
+	if err != nil {
+		return nil, fmt.Errorf("failed to marshal EJBCA Request: %w", err)
+	}
+
+	cert, err := tls.LoadX509KeyPair(e.certPath, e.keyPath)
+	if err != nil {
+		return nil, fmt.Errorf("error loading X509 key pair: %w", err)
+	}
+
+	tlsConfig := &tls.Config{
+		Certificates: []tls.Certificate{cert},
+		MinVersion:   tls.VersionTLS12,
+	}
+
+	if e.rootCAPath != "" {
+		caCert, err := os.ReadFile(e.rootCAPath)
+		if err != nil {
+			return nil, fmt.Errorf("error reading root CA: %w", err)
+		}
+		caCertPool := x509.NewCertPool()
+		caCertPool.AppendCertsFromPEM(caCert)
+		tlsConfig.RootCAs = caCertPool
+	}
+
+	client := &http.Client{Transport: &http.Transport{
+		TLSClientConfig: tlsConfig,
+	}}
+
+	req, err := http.NewRequest("POST", fmt.Sprintf("%s/ejbca-rest-api/v1/certificate/pkcs10enroll", e.serverURL), bytes.NewReader(ejbcaReqBytes))
+	if err != nil {
+		return nil, fmt.Errorf("unable to create request: %w", err)
+	}
+
+	req.Header.Set("Content-Type", "application/json")
+	req.Header.Set("Accept", "application/json")
+
+	resp, err := client.Do(req)
+	if err != nil {
+		return nil, fmt.Errorf("unable to send request: %w", err)
+	}
+
+	if resp.StatusCode != http.StatusCreated {
+		return nil, fmt.Errorf("wrong status creating certificate: %v", resp.Status)
+	}
+
+	var response Response
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return nil, fmt.Errorf("unable to read response body: %w", err)
+	}
+	if err := json.Unmarshal(body, &response); err != nil {
+		return nil, fmt.Errorf("unable to parse response body: %w", err)
+	}
+
+	// Decode certificate
+	derCert, err := base64.StdEncoding.DecodeString(response.Certificate)
+	if err != nil {
+		return nil, fmt.Errorf("unable to decode response body: %w", err)
+	}
+
+	// Decode chain
+	chain := make([]*x509.Certificate, 0)
+	for _, c := range response.CertificateChain {
+		decodedCert, err := base64.StdEncoding.DecodeString(c)
+		if err != nil {
+			return nil, fmt.Errorf("unable to decode response body: %w", err)
+		}
+		cert, err := x509.ParseCertificate(decodedCert)
+		if err != nil {
+			return nil, fmt.Errorf("unable to parse cert chain: %w", err)
+		}
+		chain = append(chain, cert)
+	}
+
+	return fulcioca.CreateCSCFromDER(derCert, chain)
```

Comment:
very nice :)

Reply:
we should revisit this in the future and thoroughly test these modules. They are important